src/backend/distributed/commands/multi_copy.c

```diff
@@ -986,6 +1165,50 @@ OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections
 
 
 /*
+ * BinaryCopyFormat iterates over columns of the relation given in copyState and
+ * looks for a column whose type is array of user-defined type. If it founds that
+ * means we cannot use binary format for COPY and it return false.
+ */
+static bool
+BinaryCopyFormat(CopyState copyState)
+{
+	TupleDesc tupleDescription = NULL;
+	Form_pg_attribute *attributes = NULL;
+	ListCell *attributeId = NULL;
+	bool binaryCopyFormat = true;
+
+	if (copyState->rel)
+	{
+		tupleDescription = copyState->rel->rd_att;
+	}
+	else
+	{
+		tupleDescription = copyState->queryDesc->tupDesc;
+	}
+
+	attributes = tupleDescription->attrs;
+	foreach(attributeId, copyState->attnumlist)
+	{
+		int attributeNumber = lfirst_int(attributeId);
+		Oid typeId = attributes[attributeNumber - 1]->atttypid;
+		bool builtInType = typeId < FirstNormalObjectId;
+		char typeCategory = '\0';
+		bool typePreferred = false;
+
+		get_type_category_preferred(typeId, &typeCategory, &typePreferred);
+
+		if (!builtInType && typeCategory == TYPCATEGORY_ARRAY)
+		{
+			binaryCopyFormat = false;
+			break;
+		}
```

Comment:
I tried this example but could not produce error;

```sql
CREATE TYPE inventory_item AS (
    name text,
    supplier_id integer,
    price numeric
);

CREATE TABLE on_hand (
  id integer,
  item inventory_item
);

SELECT master_create_distributed_table('on_hand', 'id', 'hash');

SELECT master_create_worker_shards('on_hand', 4, 1);

COPY (SELECT 1, ROW('fuzzy dice', 42, 1.99)) TO '/tmp/test';

COPY on_hand FROM '/tmp/test';
```

As far as I understand copy operation of composite type itself is made by record_send/record_recv functions, and they do not add Oid. COPY operation of array of composite type is still made by array_send/array_recv, but typCategory of that is TYPECATEGORY_ARRAY